ruby/sdk-phase-1.md

```diff
@@ -0,0 +1,439 @@
+* `Temporalio::Converter::PayloadCodec` - Unimplemented base codec
+  * Important for implementers to ensure this is async capable if using async activities anywhere
+* `Temporalio::Converter::RawValue` - Can be used as way to move around unconverted payloads
+
+#### Temporalio::Converter::PayloadConverter::JSON
+
+* Ruby has a `JSON` module in the standard library, but by default it's only primitives, arrays, hashes, etc and not
+  classes.
+  * There is a concept of "additions", but they set a `json_class` field on the object which makes the JSON not
+    cross-language compatible.
+  * There are not any good JSON-to-object conversion approaches in Ruby beyond this.
+  * Therefore, very unfortunately, it is unreasonable at this time for any area where conversion from payload to value
+    is needed to specify the type to convert to. Users can use type hints in their RBS files and such, but that has no
+    runtime check/effect.
+* The JSON encoding converter initializer will have a way to set parse and generate options.
+* The payload converter will have a way to let the default be created with JSON parse and generate options for the JSON
+  encoding converter.
+  * 💭 Why? Users wanting to simply set JSON options shouldn't have to go through the mess of recreating the default
+    list of encoding converters.
+
+## Activities
+
+A simple, heartbeating activity that calls a separate thing:
+
+```ruby
+class MyActivity < Temporalio::Activity
+  # Can customize name, e.g.
+  # activity_name 'custom-name'
+
+  # Can customize executor, e.g.
+  # activity_executor :fiber
+
+  # Can make an activity dynamic, e.g.
+  # activity_dynamic true
+
+  def initialize(db_client)
+    @db_client = db_client
+  end
+
+  def execute(arg)
+    loop do
+      db_result = @db_client.do_thing(arg)
+      if db_result
+        return db_result
+      end
+      sleep 1
+      Temporalio::Activity::Context.heartbeat
+    end
+  end
+end
+```
+
+* Activities are defined as classes.
+  * 💭 Why classes instead of methods/functions like every other SDK? There are tradeoffs to both approaches. As
+    classes, they are more easily referenced, a bit more easily typed in RBS, and metadata can be applied more easily at
+    the class level. This comes at a cost of not being able to easily share state across classes, but users can provide
+    something in the constructor multiple activities use. Open to considering activity methods instead.
+  * ❓ Would we prefer activities as methods knowing they may not be able to be typed/referenced well? We can't really
+    take the workflow signal-ish approach of those pseudo-decorators making class methods unless we say activities must
+    be instance methods (since we would allow class methods as activities).
+* Activity classes are instantiated at registration time, not for each invocation.
+  * 💭 Why? Existing Temporal Ruby SDK does instantiate for each activity task, but this discourages use of shared
```

Comment:
The temptation for programmers is to add state (in the activity object) that lingers across multiple activity invocations, something really fragile... Is there a way to freeze that activity object so that it can only be used in a pure functional style?

Reply:
Yeah, that's my concern too. I think instantiating them per-invocation is probably safer, and sharing things like dependencies could be done w/ either static fields or simply by passing them in each time.

Reply:
Technically in .NET if you use DI w/ scoped activities it does instantiate them per activity task, but overall in our SDKs you are allowed to share object/outside-of-function state across activity invocations without using globals. I am a hesitant to make Ruby the special case of forcing globals to share state.

But we can probably have ways to let the user choose. For instance, we can accept two ways of providing an activity: providing an instance and providing a class. I wouldn't want the activity class form to instantiate per attempt _by default_ because it'd be confusing to have whether you did instance vs class to have different lifetimes. But we could support some setting on the activity class to say they want it per attempt and hope they have no constructor parameters. Or we'd have to devise some DI because in general it is difficult for Temporal to control lifetimes of user activity objects due to statefulness needs.